tree/tree/src/TTreeCacheUnzip.cxx

```diff
@@ -642,58 +514,157 @@ Int_t TTreeCacheUnzip::GetRecordHeader(char *buf, Int_t maxbytes, Int_t &nbytes,
+
+   if (readbuf <= 0) {
+      fUnzipState.SetFinished(index); // Set it as not done
+      if (locbuff) delete [] locbuff;
+      return -1;
+   }
 
-      if (fUnzipStatus) delete [] fUnzipStatus;
-      if (fUnzipLen) delete [] fUnzipLen;
-      if (fUnzipChunks) delete [] fUnzipChunks;
+   GetRecordHeader(locbuff, hlen, nbytes, objlen, keylen);
 
-      fUnzipStatus  = aUnzipStatus;
-      fUnzipLen  = aUnzipLen;
-      fUnzipChunks = aUnzipChunks;
+   Int_t len = (objlen > nbytes-keylen)? keylen+objlen : nbytes;
+   // If the single unzipped chunk is really too big, reset it to not processable
+   // I.e. mark it as done but set the pointer to 0
+   // This block will be unzipped synchronously in the main thread
+   // TODO: ROOT internally breaks zipped buffers into 16MB blocks, we can probably still unzip in parallel.
+   if (len > 4*fUnzipBufferSize) {
+           if (gDebug > 0)
+                   Info("UnzipCache", "Block %d is too big, skipping.", index);
 
-      fNseekMax  = fNseek;
+           fUnzipState.SetFinished(index); // Set it as done
+           if (locbuff) delete [] locbuff;
+           return 0;
    }
 
-   fLastReadPos = 0;
-   fTotalUnzipBytes = 0;
-   fBlocksToGo = fNseek;
+   // Unzip it into a new blk
+   char *ptr = 0;
+   Int_t loclen = UnzipBuffer(&ptr, locbuff);
+   if ((loclen > 0) && (loclen == objlen+keylen)) {
+      if ( (myCycle != fCycle)  || !fIsTransferred) {
+         fUnzipState.SetFinished(index); // Set it as not done
+         if (locbuff) delete [] locbuff;
+         return 1;
+      }
+      fUnzipState.SetUnzipped(index, ptr, loclen);
+      fNUnzip++;
+   } else {
+      fUnzipState.SetFinished(index);
    }
 
-   SendUnzipStartSignal(kTRUE);
+   if (locbuff) delete [] locbuff;
+   return 0;
+}
 
+#ifdef R__USE_IMT
+////////////////////////////////////////////////////////////////////////////////
+/// We create a TTaskGroup and asynchronously maps each group of baskets(> 100 kB in total)
+/// to a task. In TTaskGroup, we use TThreadExecutor to do the actually work of unzipping 
+/// a group of basket. The purpose of creating TTaskGroup is to avoid competing with main thread.
+
+Int_t TTreeCacheUnzip::CreateTasks()
+{
+   auto mapFunction = [&]() {
+      auto unzipFunction = [&](const std::vector<Int_t> &indices) {
+         // If cache is invalidated and we should return immediately.
+         if (!fIsTransferred) return nullptr;
+
+         for (auto ii : indices) {
+            if(fUnzipState.TryUnzipping(ii)) {
+               Int_t res = UnzipCache(ii);
+               if(res)
+                  if (gDebug > 0)
+                     Info("UnzipCache", "Unzipping failed or cache is in learning state");
+            }
+         }
+         return nullptr;
+      };
+
+      Int_t accusz = 0;
+      std::vector<std::vector<Int_t>> basketIndices;
+      std::vector<Int_t> indices;
+      for (Int_t i = 0; i < fNseek; i++) {
+         while (accusz < 102400) {
```

Comment:
What is the semantic of 102400?  Why was the particular value picked?  Could the user ever want to customize this value?
Either way, please use a ```constexpr``` to record and name this value.